Validate pet set scale up/down order

dshulyak opened the pull request:
This change implements test cases to validate that:
- scale up will be done in order
- scale down in reverse order
- if any of the pets will be unhealthy, scale up/down will halt

related: https://github.com/kubernetes/kubernetes/issues/30082

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32868)

<!-- Reviewable:end -->

bprashanth commented:
Great thanks, I'll try to take a look once we have 1.4 out (O(soon))

dshulyak commented:
@bprashanth Thanks, i think it is ready for review. I will implement the last test case as a separate patch.

dshulyak commented:
@bprashanth Please review when you will have time

bprashanth commented on test/e2e/petset.go:
```diff
@@ -53,8 +55,33 @@ const (
 	// Should the test restart petset clusters?
 	// TODO: enable when we've productionzed bringup of pets in this e2e.
 	restartCluster = false
+	add            = "ADD"
```
aren't there standard api events you can use for these instead of introducing a dependency?

dshulyak replied:
i will check, i just noticed it was the approach in another test

dshulyak replied:
i wasn't able to find relevant events declared in api, do you have something specific in mind?

bprashanth commented on test/e2e/petset.go:
```diff
@@ -581,14 +721,22 @@ func (p *petSetTester) setHealthy(ps *apps.PetSet) {
 	}
 }
 
+func (p *petSetTester) setHealthy(ps *apps.PetSet) {
```
this is too easy and we know it works from all the other tests. Basically any test that invokes `saturate` does exactly this. 

Can you actually force a pet to become unhealthy? i.e it needs to fail a readiness probe or crashloop. The easiest way to do this is probably run nginx sidecar (or just as the pet) serving a file (as the readiness probe endpoint). remove the file with `kubectl exec` and it serves some non-200 code, put the file back and it's happy.

bprashanth replied:
let me know if this suggestion isn't clear and I can elaborate

dshulyak replied:
done

bprashanth replied:
I meant don't use the annotation at all to trigger "unhealthy-pet", instead: 
* bring up a petset
* trigger pet-X to fail a readiness probe (by doing the nginx idea, nginx serves file content of /tmp/foo, kubectl exec rm /tmp/foo will cause it to return 500, kubelet will probe for readiness and notice this marking the pet !ready)
* Wait till this surfaces as unready pod
* Try scaling up the petset
* Verify petset is blocked on actually unready pet 
* trigger pet-X to pass readiness probe (touch /tmp/foo)
... and so on

dshulyak replied:
by done i meant that it was clear from the first comment, i changed it in my branch and will push as soon as it will be ready

bprashanth commented on test/e2e/petset.go:
```diff
@@ -171,6 +198,119 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
+			for i, event := range Deleted(podTracker) {
+				if event.pod.Name != expectedOrder[i] {
+					framework.Failf("Order of scale down is wrong: %v != %v", expectedOrder[i], event.pod.Name)
+				}
```
can you please add one that does a scale up, scale down, scale up (without waiting for anything inbetween), and checks that we converge to the final replica count in order? (i.e going from 3->5, 5->2, 2->4 should not ever create pets 4,3 in that order)

dshulyak commented:
@bprashanth refactored a change according to discussion. I used /index.html for a readinessProbe, this way it requires less changes. Please take a look when you will have time

janetkuo commented on test/e2e/petset.go:
```diff
@@ -177,6 +209,140 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
+			close(stopCh)
+		})
+
+		It("Scale up/down in predictable order", func() {
```
nit: put more info in test descriptor, see https://github.com/kubernetes/kubernetes/blob/master/docs/devel/writing-good-e2e-tests.md#some-concrete-tools-at-your-disposal

dshulyak commented:
@janetkuo thanks for review, updated test descriptor

dshulyak commented:
@k8s-bot gci gce e2e test this

dshulyak commented:
@bprashanth @janetkuo maybe it can be merged?

janetkuo commented:
@bprashanth are you still reviewing this one?

bprashanth commented:
@janetkuo if you're ready to lgtm please go ahead

bprashanth commented on test/e2e/petset.go:
```diff
@@ -177,6 +209,140 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
+			labelSelector, err := unversioned.LabelSelectorAsSelector(&unversioned.LabelSelector{
+				MatchLabels: labels,
+			})
+			ExpectNoError(err)
```
nit: which function is this (I think there's a similar function in framework.ExpectNoError)? is there a reason it needs to be global to the e2e package if it's a petset specific thing ? (sorry this is probably my fault but I just noticed it in this pr)

bprashanth reviewed (commented):
Sorry for the delay, just had one question but LGTM (even if you don't fix the question, or want to address it but in a follow up pr or something). You do however have a rebase conflict, I think.

janetkuo commented on test/e2e/petset.go:
```diff
@@ -177,6 +209,140 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
+		var podTracker []podEvent
+
+		BeforeEach(func() {
+			framework.SkipUnlessProviderIs("gce", "vagrant")
```
Do we still need this?

janetkuo commented:
Things to consider for rebase:

1. we rename petset to statefulset and moved it to beta
2. initialized annotation behavior is flipped

janetkuo commented:
Ping @dshulyak for rebase. Would love to have this test for 1.5

dshulyak commented:
@janetkuo thanks for ping) will rebase it tomorrow in my tz

dshulyak committed:
Validate pet set scale up/down order

This change implements test cases to validate that:

    scale up will be done in order
    scale down in reverse order
    if any of the pets will be unhealthy, scale up/down will halt

Change-Id: I4487a7c9823d94a2995bbb06accdfd8f3001354c

dshulyak commented:
@janetkuo please take a look, rebased and used simple watch instead of controller

dshulyak commented on test/e2e/petset.go:
```diff
@@ -208,6 +211,112 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("verifying that scaling up before scale down is finished will not be out of order")
```
it doesn't test anything new actually, will remove

janetkuo commented on test/e2e/petset.go:
```diff
@@ -208,6 +210,82 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.StatefulSet) { ps.Spec.Replicas = 0 })
+			pst.confirmPetCount(3, ps, 10*time.Second)
+
+			By("scalling down stateful set " + psName + " to 0 replicas and waiting until none of pods will run " + ns)
```
Scaling

janetkuo replied:
will run **in namespace**

janetkuo commented on test/e2e/petset.go:
```diff
@@ -208,6 +210,82 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("creating stateful set " + psName + " in namespace " + ns)
```
Capitalize the first word of all "By" steps

janetkuo commented:
Only nits, LGTM otherwise. 

We should probably remove `framework.SkipUnlessProviderIs("gce", "vagrant")` in "Pet set recreate" test as well (this can be a follow-up PR)

janetkuo commented on test/e2e/petset.go:
```diff
@@ -208,6 +210,82 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
+			pst.breakProbe(ps, testProbe)
+			pst.waitForRunningAndNotReady(ps.Spec.Replicas, ps)
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.StatefulSet) { ps.Spec.Replicas = 3 })
+			pst.confirmPetCount(1, ps, 10*time.Second)
```
One more question: is 10s long enough?

dshulyak commented:
@janetkuo fixed comments

janetkuo commented:
/lgtm

janetkuo commented:
Thanks!
